README.md
# **BANEScarparkinglite: like BANEScarparking, but smaller!**

[![Build Status](https://travis-ci.org/owenjonesuob/BANEScarparkinglite.svg?branch=master)](https://travis-ci.org/owenjonesuob/BANEScarparkinglite)
[![codecov.io](https://codecov.io/github/owenjonesuob/BANEScarparkinglite/coverage.svg?branch=master)](https://codecov.io/github/owenjonesuob/BANEScarparkinglite?branch=master)
[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/BANEScarparkinglite)](https://cran.r-project.org/package=BANEScarparkinglite)

This is a stripped-down version of the [BANEScarparking](https://github.com/owenjonesuob/BANEScarparking) package. It contains functions for obtaining and working with the BANES car parking records from the [Bath: Hacked API](https://data.bathhacked.org/Government-and-Society/BANES-Historic-Car-Park-Occupancy/x29s-cczc), as well as some functions for getting other related datasets via web-scraping.

Feel free to contribute anything you think might be useful!

Contributors so far:

* Owen Jones ([@owenjonesuob](https://github.com/owenjonesuob))
* Ryan Kenning ([@rkenning](https://github.com/rkenning))
* Nick Howlett ([@niklz](https://github.com/niklz))

---

### **Package contents**

* `get_all_crude` (and `get_range_crude`) for retrieving the full dataset (or a subset) of records from Bath: Hacked datastore
* `refine` for processing the raw records
* `refuel` and `refuel_crude` for updating existing data frames with recent records
* `get_rugby`, `get_events` (and `get_events_detail`) and `get_daily_weather` for obtaining information through web-scraping.

The full [BANEScarparking](https://github.com/owenjonesuob/BANEScarparking) package contains all of these functions, along with parking, events, rugby and weather data up to the end of 2017 (stored as R datasets).

---

### **Installation**

You can install the package from CRAN:
```
install.packages("BANEScarparkinglite")
```

(Or if you want to keep up to date with the latest changes, you can install the development version with
```
devtools::install_github("owenjonesuob/BANEScarparkinglite")
```
You'll need the `devtools` package to do this - you can install it with `install.packages("devtools")`.)


Then you can load the package with
```
library(BANEScarparkinglite)
```
and you'll be good to go!

---

**Version:** 0.1.3 (released 2019-08-05)

**Contact:** Owen Jones (@owenjonesuob)


NEWS.md
BANEScarparkinglite v0.1.3 (Release date: 2019-08-05)
=====================================================

Bugfixes:

* Bath Hacked datastore API endpoint updated

Deletions:

* Bath Rugby's website seems to be having some trouble displaying information about results from past seasons - so `get_rugby` has been removed until something can be done about it...




BANEScarparkinglite v0.1.2 (Release date: 2018-06-30)
=====================================================

Deletions:

* Changes to the Wunderground website have left `get_daily_weather` floundering - it has been removed until a suitable alternative can be found!





BANEScarparkinglite v0.1.1 (Release date: 2018-01-20)
=====================================================

Additions:

* `get_daily_weather` now creates a weather table for any given date range (no longer limited to retrieving 398 records in one go...) and the format of this table is nicer than before (better column names etc.)
* Added code tests for all main functionality (and added `covr` reporting to Travis)
* Added NEWS file to keep track of changes

Bugfixes:

* Minor usability improvements for several functions





BANEScarparkinglite v0.1.0 (Release date: 2017-08-27)
=====================================================

Additions:

* Initial release; all functionality of main BANEScarparking package, without the attached datasets
* Travis CI is being used to check the build


inst/tests/testthat/test-web_scraping.R
context("Web scraping")


has_internet <- function() {
    !is.null(curl::nslookup("r-project.org", error = FALSE))
}



# Functions are broken so have been removed, for now

# test_that("Rugby", {
#     
#     skip_if(!has_internet())
#     
#     
#     rugby <- get_rugby(c("2015", "2016"))
#     expect_equal(nrow(rugby), 34)
#     
# })
# 
# 
# test_that("Weather", {
#     
#     skip_if(!has_internet())
#     
#     
#     weather <- get_daily_weather("2015-01-01", "2016-05-01")
#     expect_equal(nrow(weather), 
#                  as.integer(as.Date("2016-05-01") - as.Date("2015-01-01")) + 1)
# })



test_that("Events", {
    
    skip_if(!has_internet())
    
    
    event_count <- get_events("2015-01-01", "2015-03-01")
    expect_equal(nrow(event_count), 
                 as.integer(as.Date("2015-03-01") - as.Date("2015-01-01")) + 1)
    expect_is(event_count$count, "numeric")
    expect_true(all(event_count$count >= 0))
    
    
    event_details <- get_events_detail("2015-01-01", "2015-01-03")
    expect_equal(nrow(event_details), sum(event_count$count[1:3]))
})
    

R/zzz.R
.onAttach <- function(libname, pkgname) {

    messages <- vector("character", 14)
    messages[1] <- paste0("Why is R$elevation > 0?",
                          " Because R is above C level...")
    messages[2] <- paste0("Fun fact: at 15:29 on 5th May 2015, ",
                          "SouthGate General CP contained -14787 cars")
    messages[3] <- "Have you tried asking R if \"Team R\" > \"Team Python\"?"
    messages[14] <- "Bath Machine Learning Meetup => iMpenetraBLe huMan teaching"
    messages[5] <- "ALL YOUR BANES ARE BELONG TO US"
    messages[6] <- "We think, therefore we R"
    messages[7] <- "More records than Guinness"
    messages[8] <- "84% of statistics are made up on the spot"
    messages[9] <- paste0("Written by an Andrew Ng-verified \"expert on ",
                           "machine learning\"")
    messages[10] <- paste0("Putting the first thing in position 1 is just ",
                           "much more logical")
    messages[11] <- paste0("\"Error: 99 warning messages!\". *Fixes bug*. ",
                           "\"Error: 143 warning messages!\"")
    messages[12] <- paste0("All programmers are playwrights, and all computers",
                           " are lousy actors")
    messages[13] <- "There's always one more bug (Lubarsky's law)"
    messages[4] <- paste0("To understand recursion you must first ",
                           "understand recursion")

    i <- sample(1:14, 1)
    packageStartupMessage(sprintf(paste0("\nWelcome to the BANEScarparkinglite ",
                                         "package!\n\n##  %s\n"), messages[i]))
}


R/process_records.R
#' Clean up raw records
#'
#' Uses functions from the \code{dplyr} package to clean up raw records obtained
#'  from the Bath: Hacked datastore. The process is as follows:
#' \itemize{
#'  \item Select columns containing useful information only
#'  \item Remove any records with NA entries
#'  \item Remove records for "test car park"
#'  \item Convert Name and Status to factors
#'  \item Remove records with negative occupancies
#'  \item Calculate Proportion column (Occupancy/Capacity)
#'  \item Remove records with Proportion greater than \code{max_prop}
#'  \item Remove duplicate records (see \code{first_upload})
#' }
#'
#' @param x A data frame containing records to be cleaned up (e.g. the data
#'  frame obtained by calling \code{\link{get_all_crude}}).
#' @param max_prop The point at which records are discarded due to overly-full
#'  Occupancy values (default is 1.1, or 110\% full, to allow for circulating
#'  cars).
#' @param first_upload If \code{TRUE}, ensures that when duplicate records are
#'   removed, the copy which is kept is the first one uploaded after the record
#'   was taken. This takes much longer to run, due to sorting.
#' @return A data frame of clean records, with 7 columns:
#' \describe{
#'  \item{Name}{The name of the car park where the record was taken.}
#'  \item{LastUpdate}{The time the record was taken (POSIXct date-time object).}
#'  \item{DateUploaded}{The time the record was uploaded to the Bath: Hacked
#'      database (POSIXct date-time object).}
#'  \item{Occupancy}{The total number of cars in the car park.}
#'  \item{Capacity}{The number of parking spaces in the car park.}
#'  \item{Status}{Description of the change in occupancy since the previous
#'      record from that car park.}
#'  \item{Proportion}{Calculated as (Occupancy/Capacity).}
#' }
#' @examples
#' \dontshow{
#' load(system.file("tests", "testthat", "data", "raw.rda", package = "BANEScarparkinglite"))
#' refined <- refine(raw)
#' }
#' \donttest{
#' raw_data <- get_all_crude()
#' some_records <- raw_data[1:1000, ]
#'
#' dim(some_records)
#' ## 1000   16
#'
#' df <- refine(raw_data)
#' dim(df)
#' ## 813   7
#' }
#' @export

refine <- function(x, max_prop = 1.1, first_upload = FALSE) {
    # Select columns of interest
    dplyr::select(x, Name = name, LastUpdate = lastupdate,
                  DateUploaded = dateuploaded, Occupancy = occupancy,
                  Capacity = capacity, Status = status) %>%
        # Remove any records with NA entries
        stats::na.omit() %>%
        # Remove records for test car park
        dplyr::filter(Name != "test car park") %>%
        # Convert Names and Statuses to factors
        dplyr::mutate(Name = as.factor(Name), Status = as.factor(Status)) %>%
        # Remove records with negative occupancies
        dplyr::filter(Occupancy >= 0) %>%
        # Calculate Proportion column
        dplyr::mutate(Proportion = (Occupancy / Capacity)) %>%
        # Remove records with overly-full occupancies (arbitrary)
        dplyr::filter(Proportion < max_prop) %>%
        # Convert date strings to POSIXct date objects (API provides date-time
        # objects already)
        # dplyr::mutate(LastUpdate = as.POSIXct(LastUpdate, tz = "UTC",
        #                                format = "%d/%m/%Y %I:%M:%S %p"),
        #        DateUploaded = as.POSIXct(DateUploaded, tz = "UTC",
        #                                  format = "%d/%m/%Y %I:%M:%S %p")) %>%
        # Remove duplicate records:
        refine.deduplicate(first_upload = first_upload)
}


#' Remove duplicate records (internal method)
#'
#' (Internal method)
#' 
#' @param x Data frame containing records.
#' @param first_upload If TRUE, ensures record with oldest DateUploaded is kept.

refine.deduplicate <- function(x, first_upload) {
    dplyr::group_by(x, Name, LastUpdate) %>%
    {if (first_upload == FALSE) {
        # Default: Doesn't matter which we keep - only difference between
        # records is DateUploaded
        dplyr::slice(., 1L)
    } else {
        # first_upload == TRUE: ensure to choose first uploaded copy after
        # record was taken. Takes much longer due to sorting!
        dplyr::filter(., LastUpdate < DateUploaded) %>%
            dplyr::top_n(-1, DateUploaded)
    }} %>%
        dplyr::ungroup()
}


R/BANEScarparkinglite.R
#' Functions for obtaining and working with car parking data from Bath and North East Somerset
#'
#' Contains functions for importing and working with the BANES car parking
#'  records and other related datasets. For the full version of the package, including
#'  all datasets, see the repo at https://github.com/owenjonesuob/BANEScarparking.
#'
#' @docType package
#' @name BANEScarparkinglite-package
#' @author Owen Jones (\email{olj23@@bath.ac.uk})
NULL

## Quiets concerns of R CMD check re variables that appear in pipelines
if(getRversion() >= "2.15.1")  {
    utils::globalVariables(c(".", "GMT", "month", "year", "day",
                             "max_gust_kmph", "precip_type", "Date",
                             "name", "lastupdate", "dateuploaded", "occupancy",
                             "capacity", "status", "Name", "Status",
                             "Occupancy", "Capacity", "Proportion",
                             "LastUpdate", "DateUploaded"))
}


R/add_records.R
#' Download all raw records from Bath: Hacked datastore
#'
#' Reads the entire CSV file found on the Bath: Hacked datastore
#'  (\url{http://bit.ly/2i3Y1uF}). The data frame created can subsequently be
#'  updated using \code{\link{refuel_crude}}.\cr
#'  \emph{\strong{Warning:} The file is very large! This may take a while to
#'  run, depending on your internet connection.}
#'
#' @return The full dataset of car parking records
#' @examples
#' \donttest{
#' raw_data <- get_all_crude()
#'
#' str(raw_data)
#' }
#' @seealso
#' \itemize{
#'  \item \code{\link{refuel_crude}} for updating raw records
#'  \item \code{\link{refuel}} for updating data already processed with
#'   \code{\link{refine}}
#' }
#' @export

get_all_crude <- function() {
  
  token <- "npLnanLztLZey7gXvzgZ9tiS6"
  
  link <- paste0("https://data.bathhacked.org/resource/x29s-cczc.csv?",
                 "$limit=4000000&$order=dateuploaded&",
                 "$$app_token=", token)
  
  message("Making request to Socrata Open Data API (please be patient)...")
  
  df <- readr::read_csv(link, col_types = "cTcciciiiiTc")
  
  message(sprintf("Downloaded all %i records from Bath: Hacked datastore!",
                  nrow(df)))
  
  df
}


#' Add new raw records from Bath: Hacked datastore
#'
#' Update  data frame of raw records with any records that have since been added
#'  to the Bath: Hacked datastore.
#'
#' @param x The data frame of raw records from the Bath: Hacked datastore (see
#'  \code{\link{get_all_crude}}).
#' @return The data frame updated with any more recent records.
#' @examples
#' \donttest{
#' raw_data <- get_all_crude()
#' 
#' # Some time later...
#' raw_data <- refuel_crude(raw_data)
#' }
#' @seealso
#' \itemize{
#'  \item \code{\link{get_all_crude}} for obtaining data frame of raw records
#'  \item \code{\link{refuel}} for updating data already processed with
#'   \code{\link{refine}}
#' }
#' @export

refuel_crude <- function(x) {
  
  if (!identical(
    names(x),
    c("id", "lastupdate", "name", "description", "capacity", "status", 
      "occupancy", "percentage", "easting", "northing", "dateuploaded", 
      "location")
  )) {
    stop("Column mismatch! Make sure you're adding to a data frame",
         " which came from get_all_crude().")
  }
  
  token <- "npLnanLztLZey7gXvzgZ9tiS6"
  
  lastdt <- toString(max(x$dateuploaded))
  
  lastdtstr <- paste0("'", substr(lastdt, 1, 10),
                      "T", substr(lastdt, 12, 19), ".000'")
  
  new_list <- httr::GET("https://data.bathhacked.org/resource/x29s-cczc.csv",
                        query = list("$limit" = "10000000",
                                     "$order" = "dateuploaded",
                                     "$where" = paste0("dateuploaded > ",
                                                       lastdtstr),
                                     "$$app_token" = token))
  
  new_records <- httr::content(new_list, "parsed", col_types = "cTcciciiiiTc")
  
  n <- nrow(new_records)
  message(sprintf(paste0("Added %d new records from Bath: Hacked datastore!",
                         " \n  Records added: %s to %s"),
                  n, new_records$lastupdate[1], new_records$lastupdate[n]))
  
  rbind(x, new_records)
}


#' Add new processed records from Bath: Hacked datastore
#'
#' Update data frame of records already processed with \code{\link{refine}} with
#'  any records that have since been added to the Bath: Hacked datastore.
#'
#' @param x The result of calling \code{\link{refine}} on the data frame of
#'  records from the Bath: Hacked datastore (see \code{\link{get_all_crude}}).
#' @param max_prop,first_upload See \code{\link{refine}}.
#' @return The data frame updated with any more recent records.
#' @examples
#' \donttest{
#' raw_data <- get_all_crude()
#' df <- refine(get_all_crude)
#' 
#' # Some time later, add most recent records
#' df <- refuel(df)
#' }
#' @seealso
#' \itemize{
#'  \item \code{\link{get_all_crude}} for obtaining data frame of raw records
#'  \item \code{\link{refuel_crude}} for updating data frame of raw records
#' }
#' @export

refuel <- function(x, max_prop = 1.1, first_upload = FALSE) {
  if (!identical(names(x), c("Name", "LastUpdate", "DateUploaded",
                             "Occupancy", "Capacity", "Status",
                             "Proportion")))
    stop(paste0("Column mismatch! Make sure you're adding to a data frame",
                " which has been \"refined\"."))
  token <- "npLnanLztLZey7gXvzgZ9tiS6"
  lastdt <- toString(max(x$DateUploaded))
  lastdtstr <- paste0("'", substr(lastdt, 1, 10), "T", substr(lastdt, 12, 19),
                      ".000'")
  new_list <- httr::GET("https://data.bathhacked.org/resource/x29s-cczc.csv",
                        query = list("$limit" = "4000000",
                                     "$order" = "dateuploaded",
                                     "$where" = paste0("dateuploaded > ",
                                                       lastdtstr),
                                     "$$app_token" = token))
  new_records <- httr::content(new_list,
                               "parsed", col_types = "cTcciciiiiTc")
  refined <- refine(new_records, max_prop = max_prop,
                    first_upload = first_upload)
  added <- refine.deduplicate(rbind(x, refined),
                              first_upload = first_upload)
  n <- nrow(added) - nrow(x)
  message(sprintf(paste0("Refined and added %d new records from Bath: Hacked",
                         " datastore!\n  Records added: %s to %s"),
                  n, new_records$lastupdate[1], new_records$lastupdate[n]))
  added
}


#' Download records from a specified range from the Bath: Hacked datastore
#'
#' Retrieve raw records uploaded to the datastore within a specified date range and/or from a subset of car parks.
#'
#' @param from Datetime object (or "YYYY-MM-DD HH:MM:SS" string) for the earliest record to retrieve.
#' @param to Datetime object (or "YYYY-MM-DD HH:MM:SS" string) for the latest record to retrieve.
#' @param abbrs Abbreviations of names of car parks from which to retrieve
#'   records:
#'   \describe{
#'     \item{as}{Avon Street CP}
#'     \item{cs}{Charlotte Street CP}
#'     \item{l}{Lansdown P+R}
#'     \item{n}{Newbridge P+R}
#'     \item{od}{Odd Down P+R}
#'     \item{p}{Podium CP}
#'     \item{sg}{SouthGate General CP}
#'     \item{sr}{Southgate Rail CP}
#'     \item{t}{test car park}
#'   }
#' @return Car parking records from the specified date range.
#' @examples
#' # Records for 1st June 2016
#' raw_data <- get_range_crude("2016-06-01 00:00:00", "2016-06-01 23:59:59")
#' \donttest{
#' # All records from Podium CP since 14:30 on 1st January 2017
#' raw_data <- get_range_crude(from = "2017-01-01 14:30:00", abbrs = "p")
#'
#' # All records from P+Rs before 2015
#' raw_data <- get_range_crude(to = "2014-12-31 23:59:59", abbrs = c("l", "n", "od"))
#' }
#' @seealso \code{\link{get_all_crude}}
#' @export

get_range_crude <- function(from = NULL, to = NULL, abbrs = NULL) {
  
  from_dt <- lubridate::as_datetime(from)
  to_dt <- lubridate::as_datetime(to)
  
  token <- "npLnanLztLZey7gXvzgZ9tiS6"
  if (!is.null(abbrs)) {
    full <- c("Avon%20Street%20CP", "Charlotte%20Street%20CP", "Lansdown%20P%2BR",
              "Newbridge%20P%2BR", "Odd%20Down%20P%2BR", "Podium%20CP",
              "SouthGate%20General%20CP", "SouthGate%20Rail%20CP",
              "test%20car%20park")
    names(full) <- c("as", "cs", "l", "n", "od", "p", "sg", "sr", "t")
    abbrsfull <- full[abbrs]
    abbrsstring <- paste(abbrsfull, collapse = "%27,%20%27")
  }
  link <- paste0("https://data.bathhacked.org/resource/x29s-cczc.csv?",
                 "$limit=4000000&$order=dateuploaded",
                 ifelse(is.null(from), "",
                        paste0("&$where=dateuploaded%20>=%20%27",
                               gsub(" ", "T", (from_dt - 0.5)), "%27")),
                 ifelse(is.null(to), "",
                        paste0(ifelse(is.null(from), "&$where=", "%20and%20"),
                               "dateuploaded%20<=%20%27",
                               gsub(" ", "T", (to_dt - 0.5)), "%27")),
                 ifelse(is.null(abbrs), "",
                        paste0(ifelse(is.null(c(from, to)),
                                      "&$where=", "%20and%20"),
                               "name%20in(%27", abbrsstring, "%27)")),
                 "&$$app_token=", token)
  message("Making request to Socrata Open Data API...")
  df <- readr::read_csv(link, col_types = "cTcciciiiiTc")
  message(sprintf("Downloaded %i records from Bath: Hacked datastore!",
                  nrow(df)))
  df
}
